### PR TITLE
Transition history

### PR DESCRIPTION
**Pull request by szok**
Added transition history model
Added transition history table in back office asset
Added verbose_name to transition action

### CODE REVIEW

**Comment by szok** [comment]
zrzut ekranu 2015-11-04 o 12 21 38

**Comment by szok** [comment]
zrzut ekranu 2015-11-10 o 09 58 20

File: src/ralph/admin/templates/admin/templatetags/transition_history.html [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    <fieldset>
+      <legend>{% trans "Transition history" %}</legend>
+      {% if transitions_history %}
+      <table>
+        <thead>
+        <tr>
+          <th>{% trans "Date" %}</th>
+          <th>{% trans "Author" %}</th>
+          <th>{% trans "Action" %}</th>
+          <th>{% trans "Attachment" %}</th>
```
Comment (mkurek) [inline src/ralph/admin/templates/admin/templatetags/transition_history.html:14]: there should be separated column for extra data (ex. affected user) as well

Comment (szok) [inline src/ralph/admin/templates/admin/templatetags/transition_history.html:14]: Done

File: src/ralph/back_office/models.py [hunk 2]
```diff
@@ -235,6 +235,9 @@ def release_report(self, request, **kwargs):
         attachment = self._generate_report(name='release', request=request)
         attachment.description = kwargs.get('comment', '')
         attachment.save()
+
+        return attachment
```
Comment (ar4s) [inline src/ralph/back_office/models.py:273]: why? where you use returned value from this action?

Comment (szok) [inline src/ralph/back_office/models.py:273]: Write it to transition history as FK attachment.

Comment (ar4s) [inline src/ralph/back_office/models.py:273]: OK, now i see

File: src/ralph/admin/templates/admin/templatetags/transition_history.html [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                {% for value in history.data %}
+                  {{ value.0 }}: {{ value.1 }}<br />
+                {% endfor %}
+              </td>
+              <td>
+                  {% if history.attachment %}
+                    <a href="{% url history.attachment_url_name history.attachment.id history.attachment.original_filename %}">
+                      {{ history.attachment.original_filename }}
+                    </a>
+                  {% endif %}
```
Comment (ar4s) [inline src/ralph/admin/templates/admin/templatetags/transition_history.html:43]: add `{% else %}` statement with `<empty>` or something like that for better UX - what's means blank field? Error?

**Comment by ar4s** [comment]
:ok:

File: src/ralph/admin/mixins.py [hunk 4]
```diff
@@ -195,6 +195,7 @@ def changeform_view(
             extra_context['change_views'] = views
         extra_context['header_obj_name'] = self.model._meta.verbose_name
         self._initialize_search_form(extra_context)
+        extra_context['admin_model'] = self
```
Comment (mkurek) [inline src/ralph/admin/mixins.py]: i'm not sure if `admin_model` is the best name here. What do you think about `admin_view` or `admin_site`?

File: src/ralph/admin/templates/admin/templatetags/transition_history.html [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        </thead>
+        <tbody>
+          {% for history in transitions_history %}
+            <tr>
+              <td>{{ history.date }}</td>
+              <td>{{ history.user }}</td>
+              <td>{{ history.action }}</td>
+              <td>
+                {% for value in history.data %}
+                  {{ value.0 }}: {{ value.1 }}<br />
```
Comment (mkurek) [inline src/ralph/admin/templates/admin/templatetags/transition_history.html]: value could be bolded

File: src/ralph/admin/templatetags/ralph_tags.py [hunk 6]
```diff
@@ -65,3 +69,35 @@ def get_verbose_name(obj, name):
...
+
+    Example:
+        {% transition_history object %}
+    """
+    transitions_history = []
+    content_type = ContentType.objects.get_for_model(obj)
+    for history in TransitionsHistory.objects.filter(
+        transition__model__content_type=content_type, object_id=obj.pk
+    ).select_related('logged_user'):
+        for action in history.transition.actions.all():
```
Comment (mkurek) [inline src/ralph/admin/templatetags/ralph_tags.py]: you could put actions in select/prefetch related too

File: src/ralph/admin/templates/admin/templatetags/transition_history.html (hunk 1 above)
Comment (vi4m) [inline src/ralph/admin/templates/admin/templatetags/transition_history.html:14]: lerge? 

```
<div class="lerge-12 columns">
```

File: src/ralph/admin/templates/admin/templatetags/transition_history.html [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  <div class="large-12 columns">
+    <fieldset>
+      <legend>{% trans "Transition history" %}</legend>
+      {% if transitions_history %}
+      <table class="large-12 columns">
+        <thead>
+        <tr>
+          <th>{% trans "Date" %}</th>
+          <th>{% trans "Author" %}</th>
+          <th>{% trans "Action" %}</th>
```
Comment (ar4s) [inline src/ralph/admin/templates/admin/templatetags/transition_history.html]: `Actions`

File: src/ralph/lib/transitions/models.py [hunk 8]
```diff
@@ -37,8 +44,28 @@ def run_field_transition(instance, transition, field, data={}, **kwargs):
...
+                field = get_field_by_relation_path(instance, k)
+                value = v
+                if field.rel:
+                    value = str(field.rel.to.objects.get(pk=v))
+                history_kwargs[str(field.verbose_name)] = value
+
             defaults.update(kwargs)
-            func(**defaults)
+            result = func(**defaults)
+            action_names.append(str(getattr(func, 'verbose_name', '')))
```
Comment (mkurek) [inline src/ralph/lib/transitions/models.py]: if func doesn't have verbose name, (prettified) function name should be taken as default:

```
func.__name__.replace('_', ' ').capitalize()
```

File: src/ralph/admin/autocomplete.py [hunk 9]
```diff
@@ -46,11 +46,11 @@ def get(self, request, *args, **kwargs):
             {
                 'pk': obj.pk,
                 '__str__': str(obj),
-                'label': getattr(self.model, 'autocomplete_str', None),
+                'label': getattr(obj, 'autocomplete_str', None),
```
Comment (mkurek) [inline src/ralph/admin/autocomplete.py]: it's already fixed in https://github.com/allegro/ralph/pull/1894

File: src/ralph/lib/transitions/models.py [hunk 10]
```diff
@@ -37,8 +44,33 @@ def run_field_transition(instance, transition, field, data={}, **kwargs):
...
+                field = get_field_by_relation_path(instance, k)
+                value = v
+                if field.rel:
+                    value = str(field.rel.to.objects.get(pk=v))
+                history_kwargs[str(field.verbose_name)] = value
+
             defaults.update(kwargs)
-            func(**defaults)
+            result = func(**defaults)
+            print(func.__name__.replace('_', ' ').capitalize())
```
Comment (mkurek) [inline src/ralph/lib/transitions/models.py]: remove it

File: src/ralph/tests/models.py [hunk 11]
```diff
@@ -41,10 +41,12 @@ class Order(models.Model, metaclass=TransitionWorkflowBase):
     @transition_action
     def pack(self, **kwargs):
         pass
+    pack.verbose_name= 'Pack'
```
Comment (mkurek) [inline src/ralph/tests/models.py]: add translation here

**Comment by mkurek** [comment]
:ok:

**Pull request merged**